### PR TITLE
Create duty schedule

### LINKED ISSUE #3: Users

### PR DESCRIPTION
**Pull request by indocomsoft**
Closes #3.

### CODE REVIEW

File: app/controllers/duties_controller.rb [hunk 1]
```diff
@@ -8,4 +8,17 @@ def index
...
   end
+
+  def generate
+    start_date = Date.today.beginning_of_week
+    end_date = start_date + (7 * params[:num_weeks].to_i).day 
+    (start_date..end_date).each do |date| 
+      day = Date::DAYNAMES[date.wday]
+      Timeslot.where(day: day).each do |ts|
+        default_user = User.find_by(id: ts.default_user_id)
+        Duty.create(user: default_user, timeslot: ts, date: date) if !Duty.find_by(user:default_user, timeslot: ts, date: date)
```
Comment (indocomsoft) [inline app/controllers/duties_controller.rb]: use `unless` instead of `if !`, Ruby idiom

File: app/views/duties/index.html.erb [hunk 2]
```diff
@@ -2,23 +2,27 @@
...
-      <div class="schedule-timings">
-        <% @first.to_i.step(@last.to_i, 3600) do |t| %>
-          <span class="schedule-time"><%= Time.at(t).localtime.strftime('%H%M') %></span>
-        <% end %>
-      </div>
-      <ol class="schedule-days list-unstyled">
+      <table>
+        <tr>
+          <th> Days </th>
+          <% TimeRange.all.each do |tr| %>
```
Comment (indocomsoft) [inline app/views/duties/index.html.erb]: You'll need to make sure that `TimeRange` is sorted the right way

File: app/views/duties/index.html.erb [hunk 3]
```diff
@@ -2,23 +2,27 @@
...
+            <% day = Date::DAYNAMES[i % 7] %>
+            <td> <%= day %> </td>
+            <% Timeslot.where(day: day).each do |ts| %>
+              <td> <%= User.find_by(id: ts.default_user_id).email %> </td>
+            <% end %>
+          </tr>
         <% end %>
-      </ol>
+      </table>
+        <%= form_tag generate_path method: "post" %>
```
Comment (indocomsoft) [inline app/views/duties/index.html.erb]: Unclosed `form_tag` block

File: config/routes.rb [hunk 4]
```diff
@@ -5,4 +5,6 @@
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
 
   resources :duties
+  
+  post '/generate', to:  'duties#generate'
```
Comment (indocomsoft) [inline config/routes.rb]: More elegant if enclosed inside `resources :duties do` etc

File: app/controllers/duties_controller.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class DutiesController < ApplicationController
+  def index
+    @duties = Duty.all
+    tr = TimeRange.all.sort_by(&:start_time)
+    @first = tr.first.start_time
+    @last = tr.last.end_time
+    @Timeslot = Timeslot.all
```
Comment (indocomsoft) [inline app/controllers/duties_controller.rb]: Variable names are always lower-case -- only class names are Upper-case

File: app/controllers/duties_controller.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def generate
```
Comment (indocomsoft) [inline app/controllers/duties_controller.rb]: According to rubocop, Assignment Branch Condition size is too high -- need to refactor

File: app/controllers/duties_controller.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    start_date = Date.today.beginning_of_week
```
Comment (indocomsoft) [inline app/controllers/duties_controller.rb]: Use `Time.zone.today` instead

File: app/controllers/duties_controller.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def generate
+    start_date = Time.zone.today.beginning_of_week
+    end_date = start_date + (7 * params[:num_weeks].to_i).day 
+    (start_date..end_date).each do |date| 
+      day = Date::DAYNAMES[date.wday]
+      Timeslot.where(day: day).each do |ts|
+        default_user = User.find_by(id: ts.default_user_id)
+        Duty.create(user: default_user, timeslot: ts, date: date) unless Duty.find_by(user:default_user, timeslot: ts, date: date)
```
Comment (donjar) [inline app/controllers/duties_controller.rb]: `Duty.find_by_or_create`

File: app/controllers/duties_controller.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class DutiesController < ApplicationController
+  def index
+    @duties = Duty.all
+    tr = TimeRange.all.sort_by(&:start_time)
```
Comment (donjar) [inline app/controllers/duties_controller.rb]: `TimeRange.order(:start_time)`

`.all.sort_by` loads everything into Rails's memory and sorts them in place, while `.order` calls the `ORDER BY` SQL command and takes advantage of the database's representation.

File: app/controllers/duties_controller.rb (hunk 6 above)
Comment (donjar) [reply app/controllers/duties_controller.rb]: And, think about where you want to put this. Maybe in the model is better? :)

File: app/views/duties/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  <div class="col-md-12">
+    <div class="schedule-container scrollable">
+      <table>
+        <tr>
+          <th> Days </th>
+          <% TimeRange.all.each do |tr| %>
+            <th> <%= tr.start_time.strftime('%H%M') %> - <%= tr.end_time.strftime('%H%M') %> </th>
+          <% end %>
+        </tr>
+        <% (1..7).each do |i| %>
```
Comment (donjar) [inline app/views/duties/index.html.erb]: `7.times do |i|`

File: app/controllers/duties_controller.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class DutiesController < ApplicationController
+  def index
+    @duties = Duty.all
```
Comment (donjar) [inline app/controllers/duties_controller.rb]: You don't seem to use these variables at all in your views

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -3,6 +3,9 @@
   <head>
     <title>Takoyaki</title>
     <%= csrf_meta_tags %>
+    <!-- Required meta tags -->
```
Comment (donjar) [inline app/views/layouts/application.html.erb]: Separate into a new branch and PR?

And you can get rid of the comment

File: db/seeds.rb [hunk 13]
```diff
@@ -14,6 +14,8 @@
 Place.create(name: 'YIH')
 Place.create(name: 'AS8')
 
+Time.zone = 'Singapore'
```
Comment (donjar) [inline db/seeds.rb]: What is this for?

File: db/schema.rb [hunk 14]
```diff
@@ -1,69 +0,0 @@
-# This file is auto-generated from the current state of the database. Instead
```
Comment (donjar) [inline db/schema.rb]: Why is this file deleted?

Comment (donjar) [reply db/schema.rb]: Why is the Timeslot factory and Timeslots controller tests also deleted?

**Comment by donjar** [comment]
Put `.DS_Store` in your global gitignores, and don't check them in

File: app/controllers/duties_controller.rb [hunk 15]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class DutiesController < ApplicationController
+  def index
+    time_range = TimeRange.order(:start_time)
+    first = time_range.first.start_time
+    @header_iter = first.to_i.step(time_range.last.start_time.to_i, 1.hour)
+    # upper-bound exclusive, hence subtract by 1
+    @timetable_iter = first.to_i.step(time_range.last.end_time.to_i - 1, 1800)
```
Comment (donjar) [inline app/controllers/duties_controller.rb]: Change 1800

File: app/controllers/duties_controller.rb [hunk 16]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class DutiesController < ApplicationController
+  def index
```
Comment (donjar) [inline app/controllers/duties_controller.rb]: Give more descriptive names on `time_range`, `first`, `@header_iter`, `@timetable_iter`

File: app/helpers/duties_helper.rb [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+      tmp = []
+      Place.all.each do |place|
+        tmp.push(process_day_place(day, place))
+      end
+      result.push(tmp)
+    end
+    result
+  end
+
+  def process_day_place(day, place)
```
Comment (donjar) [inline app/helpers/duties_helper.rb]: If this method is only used for above, make this private

File: app/helpers/duties_helper.rb [hunk 18]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module DutiesHelper
+  def process_duties(start_date, end_date)
+    result = []
```
Comment (donjar) [inline app/helpers/duties_helper.rb]: `start_date.upto(end_date).map { |day| Place.all.map { |place| process_day_place(day, place) } }`

Will something like this work?

File: app/helpers/duties_helper.rb (hunk 17 above)
Comment (donjar) [reply app/helpers/duties_helper.rb]: And, we can talk about this later

File: app/views/duties/index.html.erb [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+                <% end %>
+                </tr>
+              <% end %>
+            <% end %>
+          </tbody>
+        </table>
+        <%= form_tag generate_path method: "post" %>
+          <%= label_tag(:num_weeks, "Generate for how many weeks ?") %>
+          <%= text_field_tag(:num_weeks) %> 
+          <%= submit_tag("Generate") %>
```
Comment (donjar) [inline app/views/duties/index.html.erb]: Mind the indentation
